### PR TITLE
Added working time counter widget.

### PR DESCRIPTION
**Pull request by areksolek**
image

### CODE REVIEW

File: modules/OSSTimeControl/dashboards/TimeCounter.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * @copyright YetiForce S.A.
+ * @license YetiForce Public License 5.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author Arkadiusz Sołek <a.solek@yetiforce.com>
+ */
+/**
+ * Time counter dashboard Class.
+ */
+class OSSTimeControl_TimeCounter_Dashboard extends Vtiger_IndexAjax_View
+{
+	public function process(App\Request $request)
```
Comment (rskrzypczak) [inline modules/OSSTimeControl/dashboards/TimeCounter.php:18]: docBlock

File: public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js [hunk 2]
```diff
@@ -3339,3 +3339,145 @@ YetiForce_Widget_Js(
...
+YetiForce_Widget_Js(
+	'YetiForce_TimeCounter_Widget_Js',
+	{},
+	{
+		hr: 0,
+		min: 0,
+		sec: 0,
+		counter: true,
+		timeStart: '',
+		timeStop: '',
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js]: docBlock

File: public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js [hunk 3]
```diff
@@ -3339,3 +3339,145 @@ YetiForce_Widget_Js(
...
+		/**
+		 * Register events on the navigation buttons.
+		 */
+		registerNavigatorButtons: function () {
+			const container = this.getContainer();
+			let instance = this;
+			let btnStart = container.find('.js-start-watch');
+			let btnStop = container.find('.js-stop-watch');
+			let btnReset = container.find('.js-reset-watch');
+			let navigator = container.find('.js-navigator');
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js]: Correction of the class name to be more specific for this widget. The current names are too general

File: public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js [hunk 4]
```diff
@@ -3339,3 +3339,145 @@ YetiForce_Widget_Js(
...
+		postLoadWidget: function () {
+			this._super();
+			this.registerNavigatorButtons();
+		},
+		/**
+		 * Register events on the navigation buttons.
+		 */
+		registerNavigatorButtons: function () {
+			const container = this.getContainer();
+			let instance = this;
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js]: remove

File: public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js [hunk 5]
```diff
@@ -3339,3 +3339,145 @@ YetiForce_Widget_Js(
...
+		 * Counting time from the moment of starting work.
+		 */
+		timeCounter: function () {
+			const container = this.getContainer();
+			if (this.counter == false) {
+				this.sec = parseInt(this.sec);
+				this.min = parseInt(this.min);
+				this.hr = parseInt(this.hr);
+				this.sec = this.sec + 1;
+				if (this.sec == 60) {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js:3462]: === in all

File: public_html/layouts/basic/styles/pages/_Dashboard.scss [hunk 6]
```diff
@@ -212,3 +212,30 @@
 		position: relative;
 	}
 }
+
+.o-time-counter {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/styles/pages/_Dashboard.scss:216]: add description

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: layouts/basic/modules/OSSTimeControl/dashboards/TimeCounterContents.tpl [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+	<div class="o-time-counter text-center">
+		<div class="o-time-counter__navigator position-relative">
+			<span class="fa-regular fa-circle o-time-counter__navigator-icon"></span>
+			<div class="o-time-counter__navigator-button js-navigator row">
+				<a class="u-cursor-pointer text-success js-start-watch" data-js="click"> <span class="fa-solid fa-circle-play"></span> </a>
+				<a class="u-cursor-pointer text-danger d-none js-stop-watch mr-1" data-js="click"> <span class="fa-solid fa-circle-stop"></span> </a>
+				<a class="u-cursor-pointer text-warning d-none js-reset-watch" data-js="click"> <span class="fa-solid fa-circle-xmark"></span> </a>
+			</div>
+		</div>
+		<div class="o-time-counter__stopwatch text-center mt-3 js-stopwatch" data-js="container">
```
Comment (rskrzypczak) [inline layouts/basic/modules/OSSTimeControl/dashboards/TimeCounterContents.tpl]: _js-stop-watch_ and _js-stopwatch_ ??

File: layouts/basic/modules/OSSTimeControl/dashboards/TimeCounter.tpl [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 5.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+	<!-- tpl-OSSTimeControl-dashboards-TimeCounter -->
+	<div class="dashboardWidgetHeader">
+		<div class="d-flex flex-row flex-nowrap no-gutters justify-content-between">
+			{include file=\App\Layout::getTemplatePath('dashboards/WidgetHeaderTitle.tpl', $MODULE_NAME) TITLE=App\Purifier::encodeHtml(App\Language::translate($WIDGET->getTitle(), 'Home'))}
```
Comment (rskrzypczak) [inline layouts/basic/modules/OSSTimeControl/dashboards/TimeCounter.tpl]: App\Language::translate($WIDGET->getTitle(), 'Home').
'Home'?

File: public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js [hunk 9]
```diff
@@ -3339,3 +3339,150 @@ YetiForce_Widget_Js(
...
+			let minutes = (dateStart.getMinutes() < 10 ? '0' : '') + dateStart.getMinutes();
+			this.timeStop = hours + ':' + minutes;
+			this.sec = 0;
+			this.min = 0;
+			this.hr = 0;
+		},
+		/**
+		 * Resets the counting operation.
+		 */
+		resetTimerCounter: function (container) {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js]: argument required? 
remove

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js [hunk 10]
```diff
@@ -3339,3 +3339,151 @@ YetiForce_Widget_Js(
 		}
 	}
 );
+YetiForce_Widget_Js(
+	'YetiForce_TimeCounter_Widget_Js',
+	{},
+	{
+		/** Hours of the timer */
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/dashboards/Widget.js]: /** @type ... */

**Pull request merged**